bin/reth/src/chain/file_client.rs

```diff
@@ -0,0 +1,118 @@
+use std::{
+    collections::HashMap,
+    path::Path,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+use reth_eth_wire::BlockBody;
+use reth_interfaces::{
+    p2p::{
+        bodies::client::{BodiesClient, BodiesFut},
+        download::DownloadClient,
+        headers::client::{HeadersClient, HeadersFut, HeadersRequest},
+        priority::Priority,
+    },
+    sync::{SyncState, SyncStateProvider, SyncStateUpdater},
+};
+use reth_primitives::{BlockHash, BlockNumber, Header, PeerId, H256};
+use tokio::{fs::File, io::BufReader};
+
+/// Front-end API for fetching chain data from a file.
+///
+/// Blocks are assumed to be written one after another in a file, as rlp bytes.
+///
+/// For example, if the file contains 3 blocks, the file is assumed to be encoded as follows:
+/// rlp(block1) || rlp(block2) || rlp(block3)
+///
+/// Blocks are assumed to have populated transactions, so reading headers will also buffer
+/// transactions in memory for use in the bodies stage.
+///
+/// Likewise, if a block body is requested and is not buffered, it will be read from the file and
+/// the header information will be buffered.
+#[derive(Debug)]
+pub struct FileClient {
+    /// The open reader for the file.
+    reader: BufReader<File>,
+
+    /// The buffered headers retrieved when fetching new bodies.
+    headers: HashMap<BlockNumber, Header>,
+
+    /// The buffered bodies retrieved when fetching new headers.
+    bodies: HashMap<BlockHash, BlockBody>,
+
+    /// Represents if we are currently syncing.
+    is_syncing: Arc<AtomicBool>,
+}
+
+impl FileClient {
+    /// Create a new file client from a file path.
+    pub async fn new<P: AsRef<Path>>(path: P) -> Result<Self, std::io::Error> {
+        let file = File::open(path).await?;
+        let reader = BufReader::new(file);
```

Comment:
Should this just be reading the entire contents of the file?

Reply:
Ideally not, but our RLP crate does not support lazily decoding the way this file is encoded - it is not an RLP list (we can lazy decode this), it is simply a big file with RLP encoded blocks after each other